File: packages/flutter-im-ui/test/swipe_reply_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/src/models/swipe_reply.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared swipe-to-reply table (`spec/swipe-reply-vectors.json`) is the one
/// rule three kits follow: which direction counts, when the list's own
/// scrolling wins instead, how far the row follows the finger, and where the
/// gesture arms. The same file is read by the SwiftUI and Compose tests, so a
/// number that drifts on one platform fails on that platform alone.
void main() {
  final table =
      jsonDecode(File('../../spec/swipe-reply-vectors.json').readAsStringSync())
          as Map<String, dynamic>;
  final cases = (table['cases'] as List).cast<Map<String, dynamic>>();

  test('the table is the geometry this kit was built with', () {
    expect(table['armDistance'], flareSwipeReplyArmDistance);
    expect(table['maxTravel'], flareSwipeReplyMaxTravel);
    expect(cases, isNotEmpty);
  });

  for (final vector in cases) {
    final id = vector['id'] as String;
    final expected = vector['expected'] as Map<String, dynamic>;
    test('swipe vector: $id', () {
      final gesture = flareSwipeReplyGesture(
        (vector['dx'] as num).toDouble(),
        (vector['dy'] as num).toDouble(),
        rtl: vector['rtl'] as bool,
      );
      expect(
        gesture.travel,
        closeTo((expected['travel'] as num).toDouble(), 0.001),
        reason: '$id travel',
      );
      expect(gesture.armed, expected['armed'], reason: '$id armed');
    });
  }
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/SwipeReplyVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.math.abs
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The shared swipe-to-reply table (`spec/swipe-reply-vectors.json`) is the one rule three kits follow:
 * which direction counts, when the list's own scrolling wins instead, how far the row follows the finger,
 * and where the gesture arms. The same file is read by the Flutter and SwiftUI tests, so a number that
 * drifts on this platform fails on this platform alone.
 */
class SwipeReplyVectorsTest {
    @Test fun theTableIsTheGeometryThisKitWasBuiltWith() {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        assertEquals(SWIPE_REPLY_ARM_DISTANCE.toDouble(), (table["armDistance"] as Number).toDouble())
        assertEquals(SWIPE_REPLY_MAX_TRAVEL.toDouble(), (table["maxTravel"] as Number).toDouble())
    }

    @Test fun everyCaseResolvesTheSameWayHere() {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        val cases = table["cases"] as List<*>
        assertTrue(cases.size >= 19, "the shared table lost cases: ${cases.size}")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val id = case["id"] as String
            val expected = case["expected"] as Map<*, *>
            val gesture = swipeReplyGesture(
                (case["dx"] as Number).toFloat(),
                (case["dy"] as Number).toFloat(),
                case["rtl"] as Boolean,
            )
            val travel = (expected["travel"] as Number).toDouble()
            assertTrue(
                abs(gesture.travel - travel) <= 0.001,
                "$id travel: expected $travel, got ${gesture.travel}",
            )
            assertEquals(expected["armed"], gesture.armed, "$id armed")
        }
    }
}

/** The table lives beside the four kits, so find the repository root from the module's directory. */
private fun vectorsFile(): File {
    var dir: File? = File(".").absoluteFile
    while (dir != null) {
        val candidate = File(dir, "spec/swipe-reply-vectors.json")
        if (candidate.isFile) return candidate
        dir = dir.parentFile
    }
    error("spec/swipe-reply-vectors.json not found above ${File(".").absolutePath}")
}
